File: 3kyu/makeASpiral.js
// 3 kyu Make a spiral
//
// Your task, is to create a NxN spiral with a given size.
// 
// For example, spiral with size 5 should look like this:
// 
// 00000
// ....0
// 000.0
// 0...0
// 00000
// 
// and with the size 10:
// 
// 0000000000
// .........0
// 00000000.0
// 0......0.0
// 0.0000.0.0
// 0.0..0.0.0
// 0.0....0.0
// 0.000000.0
// 0........0
// 0000000000
// 
// Return value should contain array of arrays, of 0 and 1, with the first row being composed of 1s. For example for given size 5 result should be:
// 
// [[1,1,1,1,1],[0,0,0,0,1],[1,1,1,0,1],[1,0,0,0,1],[1,1,1,1,1]]
// 
// Because of the edge-cases for tiny spirals, the size will be at least 5.
// 
// General rule-of-a-thumb is, that the snake made with '1' cannot touch to itself.
//
// Answer:

const spiralize = function (size) {
    const arr = Array(size).fill(0).map(e => e = Array(size).fill(0));
    let x = y = mincol = minrow = 0;
    let maxcol = arr[0].length - 1;
    let maxrow = arr.length - 1;
    let chr = 1;
    if (!maxrow && !maxcol) return arr[0];
    while (true) {
        for (x = mincol; x < maxcol; x++) arr[y][x] = chr;
        for (y = minrow; y < maxrow; y++) arr[y][x] = chr;
        for (x = maxcol; x > mincol; x--) arr[y][x] = chr;
        for (y = maxrow; y > minrow; y--) arr[y][x] = chr;
        chr = chr === 1 ? 0 : 1;
        arr[y + 1][x] = chr;

        minrow++;
        mincol++;
        maxrow--;
        maxcol--;
        x = minrow;
        y = mincol;
        if (x > maxcol || y > maxrow) break;
        if (minrow === maxrow && mincol === maxcol) {
            arr[y][x] = chr;
            break;
        }
    }
    if ((size / 2) % 2 === 0) {
        chr = chr === 1 ? 0 : 1;
        arr[y][x - 1] = chr;
    }
    return arr;
}

spiral(10);



// console.log(arr1, arr2);

// console.log(snail(arr7));


// [1,2,3,4,5,10,16,21,26,25,24,23,22,17,11,6,7,8,9,15,20,19,18,12,14]

File: 3kyu/makeASpiral.py
# 3 kyu Make a spiral
# 
# Your task, is to create a NxN spiral with a given size.
# 
# For example, spiral with size 5 should look like this:
# 
# 00000
# ....0
# 000.0
# 0...0
# 00000
# 
# and with the size 10:
# 
# 0000000000
# .........0
# 00000000.0
# 0......0.0
# 0.0000.0.0
# 0.0..0.0.0
# 0.0....0.0
# 0.000000.0
# 0........0
# 0000000000
# 
# Return value should contain array of arrays, of 0 and 1, with the first row being composed of 1s. For example for given size 5 result should be:
# 
# [[1,1,1,1,1],[0,0,0,0,1],[1,1,1,0,1],[1,0,0,0,1],[1,1,1,1,1]]
# 
# Because of the edge-cases for tiny spirals, the size will be at least 5.
# 
# General rule-of-a-thumb is, that the snake made with '1' cannot touch to itself.
# 
# Answer:
def spiralize(size):
    arr = [[0 for col in range(size)] for row in range(size)]
    chr = 1
    x = y = mincol = minrow = 0
    maxcol = len(arr[0]) - 1 
    maxrow = len(arr) - 1
    result = []
    if not maxrow and not maxcol:
        return arr[0]
    while True:
        for ax in range(mincol, maxcol):
            arr[y][x] = chr
            x += 1
        for ay in range(minrow, maxrow):
            arr[y][x] = chr
            y += 1
        for ax in range(maxcol, mincol, -1):
            arr[y][x] = chr
            x -= 1
        for ay in range(maxrow, minrow, -1):
            arr[y][x] = chr
            y -= 1
        
        chr = 0 if chr == 1 else 1
        arr[y + 1][x] = chr
        
        minrow += 1
        mincol += 1
        maxrow -= 1
        maxcol -= 1
        x = minrow
        y = mincol
        if x > maxcol or y > maxrow:
            break
        if minrow == maxrow and mincol == maxcol:
            arr[y][x] = chr
            break;

    if (size / 2) % 2 == 0:
        chr = 0 if chr == 1 else 1
        arr[y][x - 1] = chr
    return arr   